javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/JavaParserTypeDeclarationAdapter.java

```diff
@@ -68,19 +68,33 @@ public JavaParserTypeDeclarationAdapter(com.github.javaparser.ast.body.TypeDecla
         }
 
         // Look into extended classes and implemented interfaces
-        for (ResolvedReferenceType ancestor : this.typeDeclaration.getAncestors()) {
-        	try {
-	            for (ResolvedTypeDeclaration internalTypeDeclaration : ancestor.getTypeDeclaration().internalTypes()) {
-	                if (internalTypeDeclaration.getName().equals(name)) {
-	                    return SymbolReference.solved(internalTypeDeclaration);
-	                }
-	            }
-        	} catch (UnsupportedOperationException e) {
-	            // just continue using the next ancestor
+        ResolvedTypeDeclaration type = checkAncestorsForType(name, this.typeDeclaration);
+        return ((type != null) ? SymbolReference.solved(type) : context.getParent().solveType(name, typeSolver));
+    }
+
+    /**
+     * Recursively checks the ancestors of the {@param declaration} if an internal type is declared with a name equal
+     * to {@param name}.
+     * @return A ResolvedTypeDeclaration matching the {@param name}, null otherwise
+     */
+    private ResolvedTypeDeclaration checkAncestorsForType(String name, ResolvedReferenceTypeDeclaration declaration) {
+        for (ResolvedReferenceType ancestor : declaration.getAncestors()) {
```

Comment:
Should we consider the visibility of internal types?

Reply:
That check wasn't done before.
But this could be added to as method to `ResolvedReferenceTypeDeclaration` such as `getAncestors(boolean visibility)` and that method would be called here.

Reply:
well, maybe because the internal types declared in the current type are all visible, while only the non private are visible among the inherited.
Or alternatively because I was sloppy writing the original code :)